### DISCUSSION TITLE
Map skipping forward mid-song

### DISCUSSION THREAD

**Discussion by GabrielDTB** [ppy/osu#16357] [Q&A]
What I know:
Sometimes while playing a map it will 'skip' forward instantly. **This is not the standard freezing lag that I suspect would immediately come to mind.** Sometimes it fasts forwards less than a second, and sometimes multiple seconds, though I haven't been keeping track on exact times. It doesn't appear to be map specific, as I can replay a map and have no skipping. When the skip occurs, nothing else happens. The map continues seamlessly, apart from me immediately losing combo from missing after the skip. 
I'm on Linux now, but I've also had this issue in Windows 10 since at least a year ago (since my first time installing lazer).
If it would be useful, I could try to record an instance of this happening, but sometimes I can go entire days without it occurring, and other times it happens multiple times in a single play.

System Info:
OS: EndeavourOS rolling rolling (but this also happened in Windows 10)
Kernel: x86_64 Linux 5.15.12-arch1-1
Resolution: 1920x1080 (one of two 1080p monitors)
WM: i3
Disk: 842G / 5.6T (lazer is on a 3TB 5400rpm hardrive while my core system is on a separate 256GB NVMe drive. Lazer is run from the appimage release.)
CPU: AMD Ryzen 5 3600X 6-Core @ 12x 3.8GHz
GPU: AMD Radeon RX 5700 (NAVI10, DRM 3.42.0, 5.15.12-arch1-1, LLVM 13.0.0)
RAM: 32GB (I've never noticed my ram usage go above 6GB unless I manually do something that requires more ram, which never coincides with me playing osu)

Logs from a time this happened:
[database.log](https://github.com/ppy/osu/files/7832071/database.log)
[network.log](https://github.com/ppy/osu/files/7832072/network.log)
[performance.log](https://github.com/ppy/osu/files/7832073/performance.log)
[runtime.log](https://github.com/ppy/osu/files/7832074/runtime.log)

All of my recordings of the issue thus far:
[https://user-images.githubusercontent.com/61640983/148630779-1113bec5-630d-4f58-a2cd-bd400961dd62.mp4](https://user-images.githubusercontent.com/61640983/148630779-1113bec5-630d-4f58-a2cd-bd400961dd62.mp4)
[https://user-images.githubusercontent.com/61640983/148657575-648a8cb9-7d74-4318-b6db-c11d977f0a48.mp4](https://user-images.githubusercontent.com/61640983/148657575-648a8cb9-7d74-4318-b6db-c11d977f0a48.mp4)
[https://www.youtube.com/watch?v=vfJYVrHp5Yg](https://www.youtube.com/watch?v=vfJYVrHp5Yg)

**Comment by GabrielDTB** [comment]
Here is a recording of a small skip (1/8th of a second, I believe). While recording I also had files copying in the background to tax the HDD. Similar and larger skips have occurred without any load on the HDD (that I'm aware of). It seems like loading the HDD makes them more common since I got 2 skips in 15 minutes of playing. Didn't record the first one, though.
https://user-images.githubusercontent.com/61640983/148630779-1113bec5-630d-4f58-a2cd-bd400961dd62.mp4

**Reply by bdach** [reply]
can you re-record this with the frame graph showing (press ctrl-f11 twice)?

also given you're using a rarely-seen linux distro this may well depend on something in it. what's your audio setup? is it standard pulseaudio/alsa as provided by the distro, or are you running something like pipewire or jack or whichever else there is?

**Reply by GabrielDTB** [reply]
I'll try to reproduce it with the frame graph.

My audio is the standard pulseaudio setup provided by the distro. I've had this issue on Windows 10 as well, though I'll try to reproduce it there in case something changed to fix it since I last played on windows.

**Reply by GabrielDTB** [reply]
Here is a recording of it happening with the frame graph open:
https://user-images.githubusercontent.com/61640983/148657575-648a8cb9-7d74-4318-b6db-c11d977f0a48.mp4
Interestingly, this one was around 1/8th of a second as well (7 frames, while the first was 8).

**Reply by bdach** [reply]
to confirm my understanding of the issue: does the skip coincide with the time wherein the bottommost (draw) graph shows the second "red" spike?

red means swapbuffer, and a late buffer swap would definitely cause the behaviour you describe as a few rendered frames would just never make it to the actual screen. the real question is what causes that, and that may be highly dependent on your graphics card. which one do you have, and what drivers are you using with it?

**Reply by GabrielDTB** [reply]
The skip occurs 2 frames after the red spike in the draw graph. However, spikes in draw (and update) were fairly common in the full recording, which I've included below.

My GPU is a RX 5700, and I use the standard mesa drivers.
![image](https://user-images.githubusercontent.com/61640983/148663882-4a0e35f1-24e1-4c99-b355-c8cc3a9d92b5.png)

The full recording is way larger than 10MB, so here's a link to the unlisted video. I hope that's not against any rules.
[https://www.youtube.com/watch?v=vfJYVrHp5Yg](https://www.youtube.com/watch?v=vfJYVrHp5Yg)

**Reply by bdach** [reply]
not sure how to proceed further sorry. likely something specific to your setup. if it's something persistent across OSes on one machine it may even be the graphics card or the audio card.

i will say that at least for nvidia cards mesa/nouveau is inferior to the nvidia proprietary driver, but i'm not sure how that looks on the amd front.

**Reply by smoogipoo** [reply]
Are you running with unlimited frame rate? If so, can you try using 2x or 4x instead?

**Reply by GabrielDTB** [reply]
I'm running 2x refresh rate, hence the 120hz peg in the draw graph. Though, I might actually _try_ 8x or unlimited to see if anything changes.

**Reply by smoogipoo** [reply]
Right, I hadn't seen your second video above.

Very weird, I've made two observations:
1. I'm pixel peeping but I'd say that both frames after the swapbuffer are skipping. The first to a smaller degree, but the follow circle moves a bit more than I'd expect.
2. It seems like the audio also gets cut a bit. The "is" in the lyric "dark is calling" sounds much shorter than it should be.

Now, gameplay is decoupled from audio and runs on its own stopwatch, so this issue appears to be affecting at least two systems independently (update and audio, though I'd say three because the draw spike looks related), only one of which we control the timing of.
So this looks like a system timing issue and it's trying to catch up, which could also explain why the swapbuffer spike happens seemingly completely unprompted and why this happens on multiple OSs. Are there perhaps any kernel log messages around when this happens? `journalctl -k`/`dmesg -k` may reveal something.

**Reply by GabrielDTB** [reply]
Interesting. The audio always jumps forward with the video, which is how I generally notice the skips. I'll keep trying to record more instances of this occurring with the graph open to establish some commonalities between them, and I'll save the kernel logs when it happens next.

**Reply by GabrielDTB** [reply]
I've been playing on uncapped framerate and the issues seem to have abated. I also changed my fan curves, so there's the possibility for it to have been overheating issue. I can only really speculate as to what was the actual issue, but I would put more money on the uncapped framerate since I had zero overheating issues on windows and still had the skipping. Maybe my system just randomly drops frames in weird ways and with more frames I just don't notice it to record or log. Who knows.

**Reply by bdach** [reply]
there is 0 reason to run uncapped framerate unless you want to stress test your machine. the warning for that option should have made that abundantly clear.

**Reply by GabrielDTB** [reply]
Yeah, that's the weird thing. I initially tried out 8x, then tried unlimited and just forgot to switch it back. Both seemed to have eliminated skipping **and** normal dropped frames. I can't even make this stuff up lmao.

**Comment by GuineaGuy99** [comment]
Was about to create an issue about maps skipping forward mid-song, my symptoms are a little different but it may be related.

I've been playing lazer for about a year now (stable a lot longer than that), during that time I have had this bug occur around once a play session. I'm not sure if it ever has happened more than once during a session, it can take anywhere from a few minutes to a few hours before it happens, but once it does its usually fine.

Somewhat recently, I want to say around December, after a map fast forwards the game has begun showing an in-game notification stating "System audio playback is not working as expected. Some online functionality will not work. Please check your audio drivers."

Here are some examples of it happening:
https://www.youtube.com/watch?v=a4NDSl8aVqU (bug occurs at 0:46)
https://www.youtube.com/watch?v=AZgXbQThWMU (bug occurs at 0:20, I attempt to watch the replay at 2:20)

My skin has a progress bar at the bottom of the screen, it makes it easy to see the map skipping forward. ~~Interestingly in the second replay after the bug occurs the HP, score, combo and accuracy all freeze for the remainder of the map~~ (EDIT: Apparently failing was added to multiplayer as recently as yesterday? I was not aware so ignore this).

Also I am on Linux using Pulseaudio.

EDIT: I searched for existing discussions as to not create a duplicate and found this last second before I posted a new one, I saw January 8th and thought this was from last week I somehow failed to notice 2022 until just now. Sorry if that's an issue.

**Reply by peppy** [reply]
this means your system audio playback is broken and playing too slow or fast as mentioned in the message.
